conmon/ctr_logging.c

```diff
@@ -127,38 +153,67 @@ bool write_to_logs(stdpipe_t pipe, char *buf, ssize_t num_read)
 	return true;
 }
 
+
 /* write to systemd journal. If the pipe is stdout, write with notice priority,
  * otherwise, write with error priority
  */
 int write_journald(int pipe, char *buf, ssize_t buflen)
 {
-	int message_priority = LOG_INFO;
+	/* Since we know the priority values for the journal (6 being log info and 3 being log err
+	 * we can set it statically here. This will also save on runtime, at the expense of needing
+	 * to be changed if this convention is changed.
+	 */
+	char *message_priority = "PRIORITY=6";
 	if (pipe == STDERR_PIPE)
-		message_priority = LOG_ERR;
+		message_priority = "PRIORITY=3";
 
 	ptrdiff_t line_len = 0;
+
 	while (buflen > 0) {
+		writev_buffer_t bufv = {0};
+
 		bool partial = get_line_len(&line_len, buf, buflen);
-		/* sd_journal_* doesn't have an option to specify the number of bytes to write, and instead writes the
+		/* sd_journal_* doesn't have an option to specify the number of bytes to write in the message, and instead writes the
 		 * entire string. Copying every line doesn't make very much sense, so instead we do this tmp_line_end
 		 * hack to emulate separate strings.
 		 */
 		char tmp_line_end = buf[line_len];
 		buf[line_len] = '\0';
 
-		/* per docker journald logging format, CONTAINER_PARTIAL_MESSAGE is set to true if it's partial, but otherwise not set.
-		 * hence the two cases
+		/* When using writev_buffer_append_segment here, we should never approach the number of
+		 * entries necessary to flush the buffer. Therefore, the fd passed in is -1.
 		 */
-		if (partial) {
-			sd_journal_send("MESSAGE=%s", buf, "PRIORITY=%i", message_priority, "CONTAINER_ID_FULL=%s", cuuid,
-					"CONTAINER_ID=%s", short_cuuid, "CONTAINER_NAME=%s", name, "CONTAINER_PARTIAL_MESSAGE=%s", "true",
-					NULL);
-		} else {
-			sd_journal_send("MESSAGE=%s", buf, "PRIORITY=%i", message_priority, "CONTAINER_ID_FULL=%s", cuuid,
-					"CONTAINER_ID=%s", short_cuuid, "CONTAINER_NAME=%s", name, NULL);
-		}
+		_cleanup_free_ char *message = g_strdup_printf("MESSAGE=%s", buf);
+		if (writev_buffer_append_segment(-1, &bufv, message, line_len + MESSAGE_EQ_LEN) < 0)
```

Comment:
would it make sense to split the call in two:

`writev_buffer_append_segment(.., "MESSAGE=", ...)`
`writev_buffer_append_segment(.., buf, ...)`

to avoid a memory allocation?

Reply:
I'm afraid not :\ sd_journal_sendv expects the key=value to be in the same iovec. Though, I imagine a similar copy would happen with sd_journal_send, just in systemd code not ours...